delivery/media/watermarking/lib/media-delivery-watermarking.js

```diff
@@ -0,0 +1,2134 @@
+            if (clockTimestamp > v.exp + (this.wmOptions.clockTolerance || 0)) throw new Error("Token expired!");
+            return v;
+        }
+        const exp = parseInt(uint8ArrayToHex(base64url.decode(tokenParts[0], "Uint8Array")), 16);
+        if (clockTimestamp > exp + (this.wmOptions.clockTolerance || 0)) throw new Error("Token expired!");
+        const title = base64url.decode(tokenParts[1], "String"), wmid = parseInt(uint8ArrayToHex(base64url.decode(tokenParts[2], "Uint8Array")), 16), signHex = uint8ArrayToHex(base64url.decode(tokenParts[3], "Uint8Array"));
+        let isTokenVerified = !1;
+        for (const k of keys) {
+            const signature = sha256.exports.hmac(base16.decode(k, "Uint8Array"), (new TextEncoder$1).encode(`${tokenParts[0]}.${tokenParts[1]}.${tokenParts[2]}`));
+            if (uint8ArrayToHex(new Uint8Array(signature)).substring(0, 30) === signHex) {
+                isTokenVerified = !0;
+                break;
+            }
+        }
+        if (!isTokenVerified) throw new Error("Token signature verification failed!");
+        const shortToken = {
+            exp,
+            wmid,
+            title
+        };
+        return Cache.storeShortToken(cacheKey, {
+            exp,
+            wmid,
+            title
+        }), shortToken;
+    }
+    async validateToken(authToken, keys) {
+        if (!("string" == typeof authToken && !isEmptyString(authToken) || authToken instanceof Uint8Array && authToken.length > 0)) throw new Error("Invalid token type, expected non empty string or non empty Uint8Array!");
+        if (!Array.isArray(keys) || 0 == keys.length || !keys.every((item => "string" == typeof item && item.trim().length > 0))) throw new Error("Invalid keys type, expected array of hex or pem encoded strings!");
+        const wmJSON = {}, cacheKey = await buildKey([ authToken ].concat(keys)), v = Cache.getToken(cacheKey);
+        if (v) return wmJSON.header = v.header, wmJSON.payload = v.payload, this.validateClaims(wmJSON.payload), 
+        wmJSON;
+        if (this.wmOptions.tokenType === TokenType.CWT) {
+            let authTokenBuf;
+            authTokenBuf = "string" == typeof authToken ? base16.decode(authToken, "Uint8Array") : authToken;
+            const cryptoKeys = await this.importCryptoKeys(keys, authToken), cwtJSON = await this.cwtValidator.validate(authTokenBuf, cryptoKeys);
+            wmJSON.header = cwtJSON.header, wmJSON.payload = CWTUtil.claimsTranslate(Object.fromEntries(new Map(cwtJSON.payload)), claimsLabelMap);
+        } else {
+            if (this.wmOptions.tokenType !== TokenType.JWT) throw new Error("Invalid token type! Only cwt or jwt auth tokens are supported!");
+            {
+                if ("string" != typeof authToken) throw new Error("JWT token must be base64 url encoded string!");
+                const cryptoKeys = await this.importCryptoKeys(keys, authToken), jsonJWT = await this.jwtValidator.validate(authToken, cryptoKeys);
```

Comment:
If you allow the customer to import their keys and pass in the CryptoKey instances then they can choose the format. It also means less work for your module.

Reply:
We could use cryptoKeys directly if we didn't have to cache the token verification status or if we had exportKey functionality but that would also lead to overhead i guess. Thoughts?

Reply:
If the CryptoKeys persist across requests (ie, they're loaded during initialization or there is a cache of keys passed in from metadata variables), then you could use the token hash as a key into the cache. The value would be the CryptoKey. You can do an identity check on the CryptoKeys that are passed in to see if it is still present. 

Alternatively, if the CryptoKeys do not persist, you could keep a reference to the string representation of the CryptoKey as the value of the Cache.